src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,893 @@
+  return loop_info_map.at(loop_id).addr_of_write_set_var;
+}
+
+const exprt &dfcc_cfg_infot::get_outer_write_set(std::size_t loop_id) const
+{
+  PRECONDITION(is_valid_loop_id(loop_id));
+  auto outer_loop_opt = get_outer_loop_identifier(loop_id);
+  return outer_loop_opt.has_value()
+           ? get_loop_info(outer_loop_opt.value()).addr_of_write_set_var
+           : top_level_write_set;
+}
+
+const dfcc_loop_infot &
+dfcc_cfg_infot::get_loop_info(const std::size_t loop_id) const
+{
+  PRECONDITION(is_valid_loop_id(loop_id));
+  return loop_info_map.at(loop_id);
+}
+
+// find the identifier or the immediately enclosing loop in topological order
+const optionalt<std::size_t>
+dfcc_cfg_infot::get_outer_loop_identifier(const std::size_t loop_id) const
+{
+  PRECONDITION(is_valid_loop_id(loop_id));
+  auto outer_loops = get_loop_info(loop_id).outer_loops;
+
+  // find the first loop in the topological order that is connected
+  // to our node.
+  for(const auto &idx : get_loops_toposorted())
+  {
+    if(
+      std::find(outer_loops.begin(), outer_loops.end(), idx) !=
+      outer_loops.end())
+    {
+      return idx;
+    }
+  }
+  // return nullopt for loops that are not nested in other loops
+  return nullopt;
+}
+
+bool dfcc_cfg_infot::is_valid_loop_or_top_level_id(const std::size_t id) const
+{
+  return id <= loop_info_map.size();
+}
+
+bool dfcc_cfg_infot::is_valid_loop_id(const std::size_t id) const
+{
+  return id < loop_info_map.size();
+}
+
+bool dfcc_cfg_infot::is_top_level_id(const std::size_t id) const
+{
+  return id == loop_info_map.size();
+}
+
+bool dfcc_cfg_infot::must_track_decl_or_dead(
+  goto_programt::const_targett target) const
+{
+  PRECONDITION(target->is_decl() || target->is_dead());
+  auto loop_id_opt = dfcc_get_loop_id(target);
+  PRECONDITION(loop_id_opt.has_value());
+  auto loop_id = loop_id_opt.value();
+  auto &ident = target->is_decl() ? target->decl_symbol().get_identifier()
+                                  : target->dead_symbol().get_identifier();
+  if(is_top_level_id(loop_id))
+  {
+    return top_level_tracked.find(ident) != top_level_tracked.end();
+  }
+  else
+  {
+    auto &loop_info = loop_info_map.at(loop_id);
+    return loop_info.tracked.find(ident) != loop_info.tracked.end();
+  }
+}
+
+static bool __must_check_lhs(
+  const exprt &lhs,
+  const std::set<irep_idt> &local,
+  const std::set<irep_idt> &tracked,
+  messaget log,
+  const namespacet &ns)
+{
+  if(can_cast_expr<symbol_exprt>(lhs))
+  {
+    const auto &id = to_symbol_expr(lhs).get_identifier();
+
+    // direct assignments to local symbols can be skipped
+    if(local.find(id) != local.end())
+      return false;
+
+    // this is a global symbol. Ignore if it is one of the CPROVER globals
+    return !dfcc_is_cprover_symbol(id);
+  }
+  else
+  {
+    // This is a complex expression. Compute the set of root objects expressions
+    // it can assign.
+    auto root_objects = dfcc_root_objects(lhs);
```

Comment:
How about you always do this? For a symbol expression, `dfcc_root_objects` will immediately return just that symbol. I think it would reduce the code complexity here, and also make behaviour more consistent.

Reply:
This is in the else branch, so it is not a symbol expression ... What do you mean by "more consistent" ?

Reply:
I thought you could do `auto root_objects = dfcc_root_objects(lhs);` at the beginning of this function and then the `if` branch that currently exists at the top would be eliminated for `root_objects` would equally cover symbol expressions (you would have exactly one entry in `root_idents` below). I would then expect that all the use of `tracked` and `local` be consistent, which is not obvious from the code in the `if` case above: it doesn't even look at `tracked`, and I don't know why this is ok to do. Conversely, this `else` branch never does `dfcc_is_cprover_symbol`, and again I don't know why this is ok (not) to do.

Reply:
I've modified this a bit